[GPII-3965] Move log base metrics and monitoring to Terraform resources

amatas opened the pull request:
Moves all the alerts and lbm definitions to Terraform code.

This PR substitutes part of the https://github.com/gpii-ops/gpii-infra/pull/496

Before this PR is merged some manual steps must be performed in order to save the tokens for slack notification channels in prd and stg: https://github.com/gpii-ops/gpii-infra/blob/db217224558367fb91c7d0ba8d1e203834d15e20/gcp/README.md#to-receive-notifications-in-slack

natarajaya commented:
Thanks @amatas for splitting this work into smaller chunks to help us review!

Stepan have raised some good points, in addition to that, I see that you added some utility rake code to help with migration, but the whole picture is still a bit unclear to me, so I would like to see more detailed deployment / migration plan for this.

amatas commented:
> I'm missing any deployment details, does this still depend on #513? 

Yes, #513 is still needed.

> Also, @natarajaya mentioned the necessity of manual actions when recreating Slack channels?

I don't know, I have never set up a notification channel for Slack. But I will do if it is needed.

> As for the migration/cleanup process. Have you considered running the script automatically - we have a pattern for this with uptime checks: destroy_uptime_checks(["${join("\",\"", google_monitoring_uptime_check_config.this.*.name)}"]). This destroys all uptime checks, but the ones managed by TF. Perhaps doing the same for other resources would save us the necessity of manual steps?

I have tried to use the same approach, but the alerts and lbms resources can not be listed using the pattern `google_monitoring_alert_policy.*.name` . I can not find a way to get a list of the created resources.

In the case of uptime checks it is valid because the resources are created from a list using count, and they shares the same name of the resource.

> Otherwise points raised in #496 still apply:
>
>    On several occasions, there are 2 resources that only differ by stg and prd. This can be handled using TF's count construct, to avoid duplication of code and inconsistencies.

Fixed

>    Align resource naming - we have both backup-exporter and backup_exporter (for example backup_exporter_errors_prd and backup-exporter_snapshot_created_stg, also in file names btw.)

Fixed

>    Provider google-beta is defined several times, but it is used anywhere?

I don't know, it is inherited from the old TF files. Anyway I've removed in those files I found useless.

amatas commented:
> Thanks @amatas for splitting this work into smaller chunks to help us review!
> 
> Stepan have raised some good points, in addition to that, I see that you added some utility rake code to help with migration, but the whole picture is still a bit unclear to me, so I would like to see more detailed deployment / migration plan for this.

The plan is the same as before:
1. Merge #513 and promote to prd to delete old resources
1. Merge this PR to let TF create the resources again
1. If any dev environment needs to delete the resources, use the `rake clean_stackdriver_resources` task

amatas commented:
The commit https://github.com/gpii-ops/gpii-infra/pull/522/commits/87709bb5baaf383d4c77e3d6d98d0309bec4453e adds a resource that removes all the alerts, lbm and notification channels before applying the rest of the resources. The PR #513 is not needed.

amatas commented:
@natarajaya @stepanstipl can I close https://github.com/gpii-ops/gpii-infra/pull/513 in favor of the last change of this PR?

stepanstipl commented on common/modules/gcp-stackdriver/main.tf:
```diff
@@ -0,0 +1,25 @@
+provider "google-beta" {
+  project     = "${var.project_id}"
+  credentials = "${var.serviceaccount_key}"
+}
```
Still an unused `google-beta` provider here.

amatas replied:
removed

stepanstipl commented on gcp/modules/gcp-stackdriver-monitoring/main.tf:
```diff
@@ -2,109 +2,48 @@ terraform {
+  # this script is only needed to run once to migrate the resources from json to TF
   triggers = {
-    nonce = "${var.nonce}"
+    nonce = "1"
```
Looks like the triggers section is not needed at all?

amatas replied:
removed

stepanstipl commented on gcp/README.md:
```diff
@@ -449,6 +449,13 @@ See [Getting started: One-time Stackdriver Workspace setup](README.md#one-time-s
+
+If during a deployment any error is raised because either a log based metric or an alert already exist. It is possible to remove all the Stackdriver resources and let GPII-infra to recreate them.
+
+1. `rake clean_lbm_alerts`
```
This now does not match the task name -  `clean_stackdriver_resource`.

amatas replied:
fixed

natarajaya commented on gcp/README.md:
```diff
@@ -453,7 +453,7 @@ See [Getting started: One-time Stackdriver Workspace setup](README.md#one-time-s
 If during a deployment any error is raised because either a log based metric or an alert already exist. It is possible to remove all the Stackdriver resources and let GPII-infra to recreate them.
 
-1. `rake clean_lbm_alerts`
+1. `rake clean_stackdriver_resource`
```
@amatas Proper task name is `clean_stackdriver_resources`.

amatas replied:
Fixed

natarajaya commented on common/modules/gcp-stackdriver/notification_alerts_slack.tf:
```diff
@@ -0,0 +1,12 @@
+resource "google_monitoring_notification_channel" "alerts_slack" {
```
@amatas Did you try applying this? What manual steps are needed to finish configuration on Slack side? Do we need to update this README entry?
https://github.com/gpii-ops/gpii-infra/tree/master/gcp#to-receive-notifications-in-slack

amatas replied:
I can not apply this PR without a possible break of the actual resources. I will address any issue as the PR goes through the pipeline.

natarajaya replied:
Ok, I think we figured out what needs to be done to set up Slack integration properly during today's session.

natarajaya commented:
@amatas Please check couple more inline comments.  Also, there are ~100 Stackdriver primitives affected by this change, the actual contents of which impossible to review, because old json configs are discarded. How did you test this, to make sure that all LBMs, alert polices, etc are configure correctly?

amatas commented:
> > And is there still a reason for the clean_stackdriver_resources task?
> 
> If the answer is no, I would propose to drop this task.

Either way works for me, I don't have a strong opinion about leaving or deleting this utility. I only want to merge this PR and move forward with my tasks.

@natarajaya what do you think? if you say no I will drop the task, if not so I will leave this script and code as it is.

amatas commented:
> @amatas Please check couple more inline comments. Also, there are ~50 Stackdriver primitives affected by this change, the actual diff is impossible to review, because old json configs are discarded and new TF resources added. How did you test this to make sure that all LBMs, alert polices, etc are configured correctly with the new code?

I used the stackdriver.rb script to see if there is any change in the new alerts created by Terraform.

keeping in mind, that the trigger block is translated to `"trigger": null` when the values are 0 or it is not defined. The output of applying the old stackdriver provisioner over the newly created resources using TF is:

```
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "iam.modify" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "io.k8s.core.v1.pods.exec.create" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "stackdriver.alertpolicy.modify" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "backup_exporter.snapshot_created" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "gke_cluster.create" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "dns.modify" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "k8s_snapshots.error" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "compute.firewalls.modify" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "compute.instances.insert" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "backup_exporter.error" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "k8s_snapshots.couchdb.snapshot_created" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "couchdb_membership.error" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "servicemanagement.modify" is up-to-date...
null_resource.apply_stackdriver_lbm (local-exec): Log-based metric "compute.disks.createSnapshot" is up-to-date...


null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "Pod `couchdb-prometheus-exporter` exports a metric" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "IAM audit logs do not contain policy modification events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "Snapshots are being created for persistent volumes of CouchDB stateful set" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "Stackdriver audit log does not contain alert policy modification events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "Backup-exporter process does not report one or more errors" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "GKE audit log does not contain cluster creation events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "CloudDNS audit log does not contain zone modification events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "Uptime check on `flowmanager.alfredo.dev.gcp.gpii.net` is green" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "GCE audit log does not contain firewall modification events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "GCE audit log does not contain non-GKE instance creation events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "K8s cluster allocatable resource utilization stays within 85% of capacity" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "Service management log does not contain API enabling / disabling events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "K8s-snapshots logs do not contain errors" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "CouchDB membership check logs do not contain errors" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "K8s cluster API logs do not contain `kubectl exec` events" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "Backup-exporter snapshots are ok for 12 hours" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "K8s pods utilize less than 85% of persistent volume capacity" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "K8s containers does not restart more often than 1.5 times per minute" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "CouchDB request time stays within 100ms" is up-to-date...
null_resource.apply_stackdriver_monitoring (local-exec): Alert policy "K8s container CPU utilization stays within 85% of limit" is up-to-date...
```

Also this current policy has a typo in the description which is fixed in the TF resource:
https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-monitoring/resources/alert_policies/compute_instances_insert.json#L36
The `${organization_id}` variable is not interpolated in the description.

natarajaya commented:
> what do you think? if you say no I will drop the task, if not so I will leave this script and code as it is.

I think it is fine to keep the task.

> I used the stackdriver.rb script to see if there is any change in the new alerts created by Terraform.

Good approach, thanks!

So the final missing bit here is Slack integration, otherwise LGTM.

JC1738 commented on gcp/modules/gcp-stackdriver-monitoring/main.tf:
```diff
@@ -2,109 +2,45 @@ terraform {
-    enabled = "${(var.env == "prd" || var.env == "stg") ? true : false}"
-  }
+variable "crm_sa" {
+  default = "one-platform-tenant-manager@system.gserviceaccount.com"
```
What is this service account?  Where does it come from?

amatas replied:
it seems to be the Cloud Resource Manager system account. CRM uses this SA to perform some operations on the resources of a project.

natarajaya replied:
@JC1738 CRM system account can modify IAM policies of the projects on behalf of certain APIs. It is needed here, so we can exclude IAM modification events that belong to CRM account from LBM that is used by IAM modification alert policy.
